src/entry.c

```diff
@@ -261,15 +288,49 @@ entry *entryCreate(const_sds field, sds value, long long expiry) {
     bool embed_value = false;
     int embedded_field_sds_type;
     size_t expiry_size, embedded_value_sds_size, embedded_field_sds_size;
-    size_t alloc_size = entryReqSize(field, value, expiry, &embed_value, &embedded_field_sds_type, &embedded_field_sds_size, &expiry_size, &embedded_value_sds_size);
-    size_t buf_size;
+    size_t value_len = value ? sdslen(value) : SIZE_MAX;
+    size_t alloc_size = entryReqSize(sdslen(field), value_len, expiry, &embed_value, &embedded_field_sds_type, &embedded_field_sds_size, &expiry_size, &embedded_value_sds_size);
+    return entryConstruct(alloc_size, field, value, expiry, embed_value, embedded_field_sds_type, expiry_size, embedded_value_sds_size, embedded_field_sds_size);
+}
 
-    /* allocate the buffer */
-    char *buf = zmalloc_usable(alloc_size, &buf_size);
 
-    return entryWrite(buf, buf_size, field, value, expiry, embed_value, embedded_field_sds_type, embedded_field_sds_size, embedded_value_sds_size, expiry_size);
+entry *entryUpdateAsStringRef(entry *entry, const char *buf, size_t len, long long expiry) {
+    long long entry_expiry = entryGetExpiry(entry);
+    // Check for toggling expiration
+    bool expiry_add_remove = (expiry != entry_expiry) && (entry_expiry == EXPIRY_NONE || expiry == EXPIRY_NONE);
+    if (entryHasStringRef(entry) && !expiry_add_remove) {
+        stringRef *value = *entryGetValueRef(entry);
+        value->buf = buf;
+        value->len = len;
+        if (expiry != EXPIRY_NONE) *entryGetExpiryRef(entry) = expiry;
+        return entry;
+    }
+    stringRef *value = zmalloc(sizeof(stringRef));
+    value->buf = buf;
+    value->len = len;
+    if (entryHasValuePtr(entry) && !expiry_add_remove) {
+        sds *value_ref = (sds *)entryGetValueRef(entry);
+        sdsfree(*value_ref);
+        *value_ref = (sds)value;
```

Comment:
This code flow was initially confusing.  Given that the 2 IF statements are separated, it wasn't initially clear that this section is also handling the case where `expiry_add_remove` is not being changed.  Please refactor the IF structure like this for readability/maintainability.
```c
if (entryHasValuePtr(entry) && !expiry_add_remove) {
    if (entryHasStringRef(entry)) {
        // handle case for an existing stringRef
    } else {
        // handle case for an existing non-embedded sds
    }
    if (expiry != EXPIRY_NONE) *entryGetExpiryRef(entry) = expiry;
    return entry;
}
```